shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -272,16 +297,131 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+        })
+        val reverseRigid = m.toMap
+        val thisTy = TypeRef(td, td.tparams.map(p => thisCtx.targsMaps(tn.name)("")(p.name)))(prov, thisCtx)
+        thisCtx += "this" -> thisTy
+        td.mthDecls.foreach { case md @ MethodDef(rec, prt, nme, tparams, R(ty)) =>
+          implicit val prov: TypeProvenance = tp(md.toLoc, "method declaration")
+          val dummyTargs2 = tparams.map(p => freshVar(noProv/*TODO*/)(ctx.lvl + 2))
+          allTargsMaps(tn.name) += nme.name -> tparams.map(_.name).zip(dummyTargs2).toMap
+          if (!nme.name.head.isUpper)
+            err(msg"Method names must start with a capital letter", md.toLoc)
+          val bodyTy = subst(typeType(ty)(thisCtx, raise, S((tn.name, S(nme.name)))), reverseRigid)
+          decls += nme.name -> PolymorphicType(thisCtx.lvl, bodyTy)
+          thisCtx.env += "." + nme.name -> 
+            PolymorphicType(thisCtx.lvl, FunctionType(subst(thisTy, reverseRigid), bodyTy)(prov))
+        }
+        td.mthDefs.foreach { case md @ MethodDef(rec, prt, nme, tparams, L(term)) =>
+          implicit val prov: TypeProvenance = tp(md.toLoc, "method definition")
+          val dummyTargs2 = tparams.map(p => freshVar(noProv/*TODO*/)(ctx.lvl + 2))
+          allTargsMaps(tn.name) += nme.name -> tparams.map(_.name).zip(dummyTargs2).toMap
+          if (!nme.name.head.isUpper)
+            err(msg"Method names must start with a capital letter", nme.toLoc)
+          if (defs.isDefinedAt(nme.name))
+            err(msg"Method '${tn}.${nme.name}' is already defined.", nme.toLoc)
+          val bodyTy = subst(typeLetRhs(rec, nme.name, term)(thisCtx, raise, S((tn.name, S(nme.name)))), reverseRigid)
+          defs += nme.name -> bodyTy
+          thisCtx.env += "." + nme.name -> 
+            PolymorphicType(bodyTy.level, FunctionType(subst(thisTy, reverseRigid), bodyTy.body)(prov))
+        }
+        newMthDecls += tn.name -> 
+          ((td.baseClasses.flatMap(t => newMthDecls.getOrElse(t.name, Map.empty))).toMap ++ decls.toMap)
+        newMthDefs += tn.name ->
+          ((td.baseClasses.flatMap(t => newMthDefs.getOrElse(t.name, Map.empty))).toMap ++ defs.toMap)
+      }
+      ctx.copy(mthDecls = newMthDecls.toMap, mthDefs = newMthDefs.toMap)
+    } |> { implicit ctx =>
+      val newEnv = MutMap.from(ctx.env)
+      val newMthDecls = ctx.mthDecls
+      val newMthDefs = ctx.mthDefs
+      newDefs.foreach { td =>
+        implicit val prov: TypeProvenance = tp(td.toLoc, "data definition")
+        val tn = td.nme
+        val decls = newMthDecls(tn.name)
+        val defs = newMthDefs(tn.name)
+        val rigidSubsMap = ctx.getTargsMaps(td.nme.name, "").getOrElse(Map.empty).collect { 
+            case s -> (tv: TypeVariable) => tv -> freshenAbove(ctx.lvl, tv, true)
+          }.toMap[SimpleType, SimpleType]
+        val targsMap = targsMapOf(td.body)(ctx, raise, S((tn.name, N)))
+        val subsMap = td.baseClasses.flatMap(b => ctx.getTargsMaps(b.name, "")).flatten
```

Comment:
Same remark as above. Flattening collections of Map-s may override keys.

Reply:
I wrote most of this PR without having traits in mind, so I was basically assuming single inheritance, and treating `TypeDef.baseClasses` more like an option than a set. If we decide to continue supporting traits, this can be changed into a nested map, one for each class / trait.

Reply:
Addressed in 6f66ec2b249323177e94b61b3cfc5e8d8dc43015